### PR TITLE
Templating type definition

### PR DESCRIPTION
**Pull request by NoHomey**
case 2. Improvement to existing type definition.

@paulvanbrenk Labeled #10694 with [Revision requested](https://github.com/DefinitelyTyped/DefinitelyTyped/labels/Revision%20requested) which also exports function like Object.assign. Object.assign is templated in [lib.es6.d.ts Object.assign](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es6.d.ts#L4399) which make sense to change function exported by 'object-assign' to  generic as well.

### CODE REVIEW

**Comment by paulvanbrenk** [comment]
THis looks good to me. @chbrown can you take a look too?

File: object-assign/object-assign.d.ts [hunk 1]
```diff
@@ -4,6 +4,11 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 declare module "object-assign" {
+  function objectAssign<T, U>(target: T, source: U): T & U;
+  function objectAssign<T, U, V>(target: T, source1: U, source2: V): T & U & V;
+  function objectAssign<T, U, V, W>(target: T, source1: U, source2: V, source3: W): T & U & V & W;
+  function objectAssign<T, U, V, W, Q>(target: T, source1: U, source2: V, source3: W, source4: Q): T & U & V & W & Q;
+  function objectAssign<T, U, V, W, Q, R>(target: T, source1: U, source2: V, source3: W, source4: Q, source5: T): T & U & V & W & Q & R;
```
Comment (chbrown) [inline object-assign/object-assign.d.ts]: source5 should have type R, not T.

**Comment by chbrown** [comment]
I pointed out a typo in my comment above. Copy & paste errors like that are a good reason for tests, so if you could add tests, that'd be great.

I would also suggest moving the react-redux changes to a different PR. 

Otherwise looks fine and matches Object.assign in lib.es6.d.ts as specified, so that's a good sign.

**Comment by NoHomey** [comment]
I've added tests to test the new generic type definition and moved react-redux changes to #10727.

**Comment by rbuckton** [comment]
@NoHomey looks good. Thank you for the contribution.

**Pull request merged**